File: TotalCrossVM/src/launchers/darwin/Launcher.m
/*********************************************************************************
 *  TotalCross Virtual Machine, version 1                                        *
 *  Copyright (C) 2007-2012 SuperWaba Ltda.                                      *
 *  All Rights Reserved                                                          *
 *********************************************************************************/


#if HAVE_CONFIG_H
#include "config.h"
#endif

#define AUDIO_COMPILE_ISSUE

#import <UIKit/UIKit.h>
#import <Foundation/NSThread.h>
#include <stdio.h>
#include <dlfcn.h>
#import <UIKit/UIAlert.h>

/* undef rpl_realloc */
#undef realloc

typedef id Context;

typedef int  (*StartVMProc)             (char* args, Context context);
typedef void (*NotifyStopVMProc)        ();
typedef int  (*StartProgramProc)        (Context context);

typedef void *dlHandle;

static char *cmdLine;

@interface LauncherMain : NSObject <UIApplicationDelegate>
{
   int startupRC;
   StartProgramProc fStartProgram;
   dlHandle tcvm;
   Context context;
}
- (void)   mainLoop: (id)param;
- (float)  systemVolume;
- (void) fatalError: (NSString*)msg;
@end

@implementation LauncherMain

-(id) init
{
	[super init];
    fStartProgram = NULL;
	return self;
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
{
   [ alertView release ];
   [[UIApplication sharedApplication] terminate];
}

- (void) fatalError: (NSString*)msg
{
   UIAlertView *myAlert = [ [ UIAlertView alloc ]
      initWithTitle:@"Fatal Error"
      message: msg
      delegate:self
      cancelButtonTitle: nil
      otherButtonTitles: @"Quit", nil];
   [ myAlert show ];
}

- (void) mainLoop: (id)param
{
   [[NSAutoreleasePool alloc] init];

   if (startupRC == 0)
   {
      fStartProgram = (StartProgramProc)dlsym(tcvm, "startProgram");
      if (!fStartProgram)
      {
         [ self fatalError: @"Cannot find the 'startProgram' entry point" ];
         return;
      }
      fStartProgram(context);
   }

   dlclose(tcvm); // free the library
   free(cmdLine);
   [[UIApplication sharedApplication] terminate];
}

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification
{
   // setup for device orientation change events
   [[ UIDevice currentDevice ] beginGeneratingDeviceOrientationNotifications ];

   tcvm = dlopen("libtcvm.dylib", RTLD_LAZY);                    // load in current folder - otherwise, we'll not be able to debug
   if (!tcvm)
      tcvm = dlopen("../libtcvm.dylib", RTLD_LAZY);              // load in parent folder
   if (!tcvm)
      tcvm = dlopen("/Applications/TotalCross.app/libtcvm.dylib", RTLD_LAZY); // load in most common absolute path
   if (!tcvm)
   {
      [ self fatalError: @"Cannot find the 'libtcvm' shared lib" ];
      return;
   }

   StartVMProc fStartVM = (StartVMProc)dlsym(tcvm, "startVM");
   if (!fStartVM)
   {
      [ self fatalError: @"Cannot find the 'startVM' entry point" ];
      return;
   }

   startupRC = fStartVM(cmdLine, &context);

   //printf("volume = %f\n", [self systemVolume]);
   [NSThread detachNewThreadSelector:@selector(mainLoop:) toTarget:self withObject:nil];
}

- (void)applicationWillTerminate:(UIApplication *)application
{
   if ([NSThread isMainThread])
   {
      NotifyStopVMProc fNotifyStopVM = (NotifyStopVMProc)dlsym(tcvm, "notifyStopVM");
      if (!fNotifyStopVM)
      {
         [ self fatalError: @"Cannot find the 'notifyStopVM' entry point" ];
         return;
      }
      fNotifyStopVM();

      [ NSThread sleepForTimeInterval: 1800.0 ]; // wait a maximum of 30 minutes, once notified the VM thread should end before 
   }
}

// Getting volume
- (float)systemVolume
{
    float         volume = 0, volumeL, volumeR;
#ifndef AUDIO_COMPILE_ISSUE
    OSStatus      err;
    AudioDeviceID device;
    UInt32        size;
    UInt32        channels[2];

    size = sizeof device;
    err = AudioHardwareGetProperty(kAudioHardwarePropertyDefaultOutputDevice, &size, &device);

    size = sizeof channels;
    if (noErr == err) err = AudioDeviceGetProperty(device, 0, false, kAudioDevicePropertyPreferredChannelsForStereo, &size, &channels);

    size = sizeof volume;
    if (noErr == err) err = AudioDeviceGetProperty(device, channels[0], false, kAudioDevicePropertyVolumeScalar, &size, &volumeL);
    if (noErr == err) err = AudioDeviceGetProperty(device, channels[1], false, kAudioDevicePropertyVolumeScalar, &size, &volumeR);

    if (noErr == err)
    {
        // Select greatest volume
        volume = (volumeL < volumeR) ? volumeR : volumeL;
    }
    else
    {
        volume = 1.0;	// Fallback
    }
#endif
    return volume;
}

//- (void)setView: (TCView*)view
//{
//   _currentView = view;
//}

@end

#define LAUNCHED_FROM_SPRINGBOARD "--launchedFromSB"

int main(int argc, char *argv[])
{
   cmdLine = strdup(argv && argv[0] ? argv[0] : "");
   if (argc > 1) // if there's a commandline passed by the system or one passed by the user
   {
      cmdLine = (char*)realloc(cmdLine, strlen(cmdLine) + sizeof(" /cmd "));
      strcat(cmdLine, " /cmd ");
      char **p = argv + 1;
      int n = argc;
      while (n-- > 1)
      {
         // remove the "launched by the SpringBoard" option
         if (strcmp(*p, LAUNCHED_FROM_SPRINGBOARD))
         {
            cmdLine = (char*)realloc(cmdLine, strlen(cmdLine) + strlen(*p) + 2);
            strcat(cmdLine, " ");
            strcat(cmdLine, *p);
         }
         p++;
      }
   }

   [[NSAutoreleasePool alloc] init];
   return UIApplicationMain(argc, argv, nil, @"LauncherMain");
}

File: TotalCrossVM/src/launchers/linux/Launcher.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2021 TotalCross Global Mobile Platform Ltda.
// Copyright (C) 2022-2026 Amalgam Solucoes em TI Ltda
//
// SPDX-License-Identifier: LGPL-2.1-only

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <dlfcn.h>
#include <string.h>
#include <stdio.h>
#include "xtypes.h"

#if __APPLE__
#define CURRENT_PATH "@executable_path"
#define PARENT_PATH "@executable_path/.."
#define LIB_EXTENSION "dylib"
#else
#define CURRENT_PATH "."
#define PARENT_PATH ".."
#define LIB_EXTENSION "so"
#endif

char *args = "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890";

typedef int (*ExecuteProgramProc)(char* args);

typedef void *Handle;

static Handle tryOpen(const char *prefix, const char* libname)
{
   char path[MAX_PATHNAME];

   if (prefix) {
      snprintf(path, MAX_PATHNAME, "%s/%s.%s", prefix, libname, LIB_EXTENSION);
   }
   else {
      snprintf(path, MAX_PATHNAME, "%s.%s", libname, LIB_EXTENSION);
   }
   return dlopen(path, RTLD_LAZY);
}

static int executeProgram(char* cmdline)
{
   int ret = 0;
   ExecuteProgramProc fExecuteProgram = NULL;
   Handle tcvm;

   tcvm = tryOpen(CURRENT_PATH, "libtcvm");              // load in current folder - otherwise, we'll not be able to debug

   if (!tcvm) {
      printf("%s\n", dlerror());
      tcvm = tryOpen(PARENT_PATH, "libtcvm");            // load in parent folder
   }
#if __APPLE__
   if (!tcvm) {
      printf("%s\n", dlerror());
      tcvm = tryOpen("@rpath", "libtcvm");               // LC_RPATH pointing to a predefined path, like /usr/local/lib/totalcross
   }
   if (!tcvm) {
      printf("%s\n", dlerror());
      tcvm = tryOpen(NULL, "libtcvm");                   // search in paths from loader
   }
#endif
   if (!tcvm) {
      printf("%s\n", dlerror());
      tcvm = tryOpen("/usr/lib/totalcross", "libtcvm");  // load in most common absolute path
   }
   if (!tcvm) {
      printf("%s\n", dlerror());
      return 10000;
   }
   fExecuteProgram = (ExecuteProgramProc)dlsym(tcvm, TEXT("executeProgram"));
   if (!fExecuteProgram)
      return 10001;

   ret = fExecuteProgram(cmdline); // call the function now

   dlclose(tcvm); // free the library
   return ret;
}

int main(int argc, const char *argv[])
{
   char cmdline[512];
   xmemzero(cmdline,sizeof(cmdline));
   int argvIndex = 0;
   if (argv)
   {
      if (argc > 1 && 
         xstrlen(argv[0]) >= 8 &&
         memcmp(argv[0] + xstrlen(argv[0]) - 8, "Launcher", 8) == 0) {
         argvIndex = 1;
      }
      xstrcpy(cmdline, argv[argvIndex++]);
      xstrcat(cmdline, ".tcz");
   }
   if (argc > argvIndex || args[0] != '1') // if there's a commandline passed by the system or one passed by the user
   {
      xstrcat(cmdline, " /cmd ");
      if (args[0] != '1')
         xstrcat(cmdline, args);
      const char **p = argv + argvIndex;
      int n = argc;
      while (n-- > argvIndex)
      {
         xstrcat(cmdline, " ");
         xstrcat(cmdline, *p++);
      }
   }
   return executeProgram(cmdline); // in tcvm\startup.c
}
